File: findTiesAndAccents.ts
inlets = 2;
outlets = 2;

let floats: number[] = [];

let choices: string[] = [];

let bot: number;
let top: number;


function list() {
    switch (inlet) {
        case 0:
            floats = arrayfromargs(arguments);
            bang();
            break;
    }
}

function setChoices() {
    choices = arrayfromargs(arguments);
    bang();
}

function setBot(n: number) {
    bot = n;
    bang();
}

function setTop(n: number) {
    top = n;
    bang();
}

function bang() {
    const stages = [
        bot,
        1 + top,
        1
    ]
    const durations: number[] = []
    const velocities: number[] = []
    const addChoice = (choice: string | undefined) => {
        if (choice !== undefined) {
            velocities.push(
            choice.indexOf('R') >= 0
                ? 0
                : choice.indexOf('A') >= 0
                    ? 127
                    : 63
        )
        durations.push(
            choice.indexOf('T') >= 0
                ? 120
                : 60
            )
        }
    }
    for (let f of floats) {
        let i: number
        f = f * nextFloatDown;
        for (i = 0; i < stages.length; i++) {
            if (f < stages[i]) {
                break;
            }
        }
        addChoice(choices[i]);
    }
    if (durations.length > 0 && velocities.length > 0) {
        outlet(0, ["duration", 1, ...durations]);
        outlet(0, ["velocity", 1, ...velocities]);
        outlet(1, velocities.map(x => +(x === 127)));
    }
}

const nextFloatDown = (1 - 2 ** -53);

File: findTiesAndAccents.java
import com.cycling74.max.*;

public class findTiesAndAccents extends MaxObject {
    Atom[] atoms = new Atom[0];
    String[] choices = new String[0];
    float bot = 0, top = 0;
    int nudge = 0;

    findTiesAndAccents() {
        declareInlets(new int[] { DataTypes.LIST, DataTypes.MESSAGE });
        declareOutlets(new int[] { DataTypes.MESSAGE });
    }

    protected void list(Atom[] a) {
        final int idx = getInlet();
        switch (idx) {
            case 0:
                atoms = a;
                bang();
                break;
        }
    }

    public void setChoices(Atom[] a) {
        choices = Atom.toString(a);
        bang();
    }

    public void setBot(float n) {
        bot = n;
        bang();
    }

    public void setTop(float n) {
        top = n;
        bang();
    }

    public void setNudge(int n) {
        nudge = n;
        bang();
    }

    protected void bang() {
        if (choices != null && choices.length > 0 && atoms != null && atoms.length > 0) {
            final float[] stages = new float[] { bot, 1 + top, 1 };
            final float[] floats = Atom.toFloat(Atom.rotate(atoms, nudge));

            int[] durations = new int[floats.length + 1];
            durations[0] = 1;
            int[] velocities = new int[floats.length + 1];
            velocities[0] = 1;
            int[] accents = new int[floats.length];

            for (int i = 0; i < floats.length; i++) {
                int j;
                float f = Math.nextDown(floats[i]);
                for (j = 0; j < stages.length; j++) {
                    if (f < stages[j]) {
                        break;
                    }
                }

                String choice = choices[j];

                velocities[i + 1] = choice.contains("R") ? 0 : choice.contains("A") ? 127 : 63;
                accents[i] = choice.contains("A") ? 1 : 0;
                durations[i + 1] = choice.contains("T") ? 120 : 60;
            }

            outlet(0, "duration", durations);
            outlet(0, "velocity", velocities);
            outlet(1, accents);
        }
    }
}
